src/server/frontend_wayland/wl_pointer.cpp

```diff
@@ -212,18 +281,53 @@ void mf::WlPointer::release()
 }
 
 WlStreamCursor::WlStreamCursor(
-    std::shared_ptr<mf::BufferStream> const& stream,
+    std::shared_ptr<mc::BufferStream> const& stream,
     geom::Displacement hotspot) :
     stream{stream},
     hotspot{hotspot}
 {
+    stream->set_frame_posted_callback(
+        [this](auto)
+        {
+            this->apply_latest_buffer();
+        });
+}
+
+WlStreamCursor::~WlStreamCursor()
+{
+    stream->set_frame_posted_callback([](auto){});
 }
 
 void WlStreamCursor::apply_to(mf::WlSurface* surface)
 {
-    if (auto scene_surface = surface->scene_surface())
+    auto const scene_surface = surface->scene_surface();
+
+    if (scene_surface)
     {
-        scene_surface.value()->set_cursor_stream(stream, hotspot);
+        surface_under_cursor = *scene_surface;
+        apply_latest_buffer();
+    }
+    else
+    {
+        surface_under_cursor.reset();
+    }
+}
+
+void WlStreamCursor::apply_latest_buffer()
+{
+    if (auto const surface = surface_under_cursor.lock())
+    {
+        if (stream->has_submitted_buffer())
+        {
+            auto const cursor_image = std::make_shared<BufferCursorImage>(
```

Comment:
This is an additional memory allocation per frame than #1034, but this path is far from optimal anyway (and we can fix this if it ever shows up on a performance metric).